### ISSUE TITLE
Firebird, DatePart SECOND/MILLISECOND

### ISSUE THREAD

**Issue by dmitry-lipetsk** [linq2db/linq2db#967]
### Environment details
linq2db version: *Master*  
Database Server: *Firebird*  
Database Provider: *Firebird*  

Firebird data provider translates DatePart for Second and Millesecond to

EXTRACT(SECOND FROM <timestamp>) - returns NUMERIC(9,4)
EXTRACT(MILLISECOND FROM <timestamp>) - returns NUMERIC(9,1)

As I understand, linq2db expected INTEGER result, not NUMERIC with fraction

As a result, Issue882Tests disables tests 'Second' and 'Millisecond' for Firebird.

These problems can be resolved through modifiction of DataPart template for Firebird

work\source\linqtodb\sql\sql.cs
```
[Sql.DatePart(PN.Firebird, "TRUNC(Extract({0} from {{0}}))", ...]
public static int? DatePart(DateParts part, DateTime? date)
```

**Comment by dmitry-lipetsk** [comment]
This modification will resolve the problem with SECOND/MILLISECOND, but breaks down work of DayOfYear/Quarter/WeekDay functions.

So, need use TRUNC only for SECOND and MILLISECOND parts :)

Or, correct code in FirebirdSqlOptmizer.cs
```
			else if (expr is SqlExpression)
			{
				SqlExpression e = (SqlExpression)expr;

				if (e.Expr.StartsWith("TRUNC(Extract(Quarter"))
					return Inc(Div(Dec(new SqlExpression(e.SystemType, "Extract(Month from {0})", e.Parameters)), 3));

				if (e.Expr.StartsWith("TRUNC(Extract(YearDay"))
					return Inc(new SqlExpression(e.SystemType, e.Expr.Replace("Extract(YearDay", "Extract(yearDay"), e.Parameters));

				if (e.Expr.StartsWith("TRUNC(Extract(WeekDay"))
					return Inc(new SqlExpression(e.SystemType, e.Expr.Replace("Extract(WeekDay", "Extract(weekDay"), e.Parameters));
			}

```

**Comment by sdanyliv** [comment]
@dmitry-lipetsk, it is better in thi situation to make PR.

**Comment by dmitry-lipetsk** [comment]
I not sure that is completed decision.

May be, after TRUNC, need also made a CAST to INTEGER :)

**Comment by sdanyliv** [comment]
Fixed in 4ca152b4a53a7e85589d0c9071062fd7a5784681
